✨ [Feature] 사용자 회원가입 및 로그인 API

joosomi opened the pull request:
## Summary
회원가입 : `POST /api/members/sign-up`
로그인: `POST /api/members/sign-in`
새 액세스 토큰 발급 API: `POST /api/members/auth/refresh`

## Describe your changes
- 로그인할 경우에는 액세스 토큰과 리프레쉬 토큰이 발급되고 리프레쉬 토큰은 레디스에 저장됩니다.(ttl: 30일)
- `@RefreshAuth` 데코레이터를 만들어서 전역인 액세스 토큰 검증을 거치지 않고, 리프레쉬 토큰 검증 guard만 거치도록 일단 이렇게 설정했습니다. 
-> 리프레쉬 토큰 검증 후 -> 사용자 id를 request 객체에 저장 -> 이 id를 사용하여 db에서 사용자 정보(`id`, `accountName`, `name`)를 가져와서 새로운 payload를 담아 새로운 액세스 토큰이 발급됨 
(-> 현재는 회원 정보 수정하는 로직이 없지만, 실제로는 payload가 바뀔 가능성도 있기 때문에 새로운 payload를 담음)

## Issue number and link
회원가입 #15 
로그인/토큰 발급 #16

joosomi committed:
✨ feat: 사용자 회원가입: 회원 정보 저장 및 중복 계정 처리(#15)

joosomi committed:
✨ feat: 사용자 로그인: 액세스, 리프레쉬 토큰 생성 (#16)

joosomi committed:
♻️ refactor: 반환 타입 명시, HTTP 상태 코드 추가 및 import 구문 정리

joosomi committed:
✨ feat: redis cache 설정 및 리프레쉬 토큰 저장 로직(#16)

joosomi committed:
✨feat: redis docker 환경 설정 (#16)

joosomi committed:
✨ feat: jwt authguard 전역 설정 및 public decorator 설정(#16)

joosomi committed:
♻️ refactor: JwtAuthGuard에 제네릭 타입 추가 (#16)

- lint 규칙 체크하여 수정
- handleRequest 메서드에 제네릭 타입 TUser를 추가하여 타입 안전성과 유연성 향상.
- canActivate 메서드의 반환 타입을 명시적으로 정의.

joosomi committed:
✨ feat: dev guard 생성 및 request에 user 정보 할당(#16)

jis-kim commented on docker-compose.dev.yml:
```diff
@@ -16,6 +16,15 @@ services:
       timeout: 5s
       retries: 5
+  redis:
+    image: 'redis:latest'
```
버전을 지정해주셨으면 좋겠습니다! 이유: 나중에 redis 버전 업으로 latest 버전이 바뀌면 docker 를 사용해서 환경을 일정하게 하는 이유가 없다고 생각

joosomi commented on src/auth/strategies/jwt.types.ts:
```diff
@@ -0,0 +1,12 @@
+// 향후 확장 가능성(예: 추가 필드)을 고려할 때 interface 선언
```
type으로 수정 ( 이유 : 통일성 유지 그리고 확장 가능성 없어보임)

jis-kim commented on src/auth/guards/jwt-auth.guard.ts:
```diff
@@ -0,0 +1,41 @@
+
+  // TUser 제네릭을 사용하여 기본값을 JwtUser로 설정
+  // JwtUser 타입을 기본으로 사용하면서도 필요에 따라 다른 타입 지원 가능
+  handleRequest<TUser extends JwtUser = JwtUser>(
```
이거 `AuthGuard('jwt'`)의 함수를 override 한 것 같은데, 제 생각에는 `AuthGuard('jwt')`의 `handleRequest` 와 동일한 역할을 하는 것 같아 정의가 필요없을 것 같습니다!
확실하지는 않아 더 알아보고 같이 확인해보면 좋을 것 같아요.

joosomi committed:
🐛 fix: mockuser id uuid 형식으로 수정

joosomi committed:
♻️ refactor: 도커 redis 버전 7.4.0으로 고정

joosomi committed:
♻️ refactor: Payload, User interface에서 type으로 통일

joosomi committed:
♻️ refactor: 토큰 예외 처리 로직 - 토큰 만료, invalid token

joosomi committed:
✨ feat: JWT 리프레쉬 토큰 로직 (#16)

- 새로운 액세스 토큰 발급 API 구현
- 리프레시 토큰 엔드포인트를 'auth/refresh' 로 memberscontroller에 추가
- 하지만 refreshtoken 검증 guard 생성 필요

joosomi committed:
♻️ refactor: handleRequest 메서드 타입 지정(#16)

- handleRequest 메서드에 제네릭 타입 TUser와 AuthInfo 타입을 사용

joosomi committed:
✨ feat: RefreshToken guard 생성 및 토큰 예외 처리 추가 (#16)

joosomi committed:
♻️ refactor: 컨트롤러에서 불필요한 useGuards 제거

- refreshAuth 데코레이터에서 jwtRefreshGuard를 이미 사용
- 굳이 controller에서 useGuard를 선언해 줄 필요 없었음

jis-kim commented on src/auth/auth.service.ts:
```diff
@@ -0,0 +1,103 @@
+
+    //비밀번호 검증
+    const isPasswordValid = await bcrypt.compare(password, member.password);
+    if (isPasswordValid == false) {
```
```suggestion
    if (isPasswordValid === false) {
```
여기 `=`이 하나 빠졌습니당

jis-kim commented on src/auth/auth.service.ts:
```diff
@@ -0,0 +1,103 @@
+
+    const payload = { sub: member.id, accountName: member.accountName, name: member.name };
+    const newAccessToken = this.jwtService.sign(payload, { expiresIn: JWT_ACCESS_EXPIRES_IN });
+    return { newAccessToken };
```
`newAccessToken` 보다 로그인 할 때랑 똑같이 `accessToken`으로 주는 게 클라이언트에서 처리하기 편하지 않을까요...? 개인 생각입니다!

joosomi committed:
♻️ refactor: == 대신 === 사용하여 비밀번호 검증, 변수명 일관성 위해 수정

joosomi committed:
✅ test: 테스트 코드 작성 (#15, #16)

jis-kim reviewed (approved):
고생 많으셨습니다! 👍

ooheunda reviewed (approved):
정말 고생하셨습니다~! 👍

joosomi committed:
♻️ refactor: ReviewRepository mock으로 제공
